go/private/actions/compilepkg.bzl

```diff
@@ -95,8 +95,10 @@ def emit_compilepkg(
         archives = archives + [go.coverdata]
 
     sdk = go.sdk
-    inputs_direct = (sources + embedsrcs + [sdk.package_list] +
-                     [archive.data.export_file for archive in archives])
+    inputs_direct = sources + embedsrcs + [sdk.package_list]
+    for archive in archives:
+        inputs_direct.append(archive.data.export_file)
+        inputs_direct.extend(archive.data._headers)
```

Comment:
It looks like this will add headers from direct dependencies, but you also need headers from transitive dependencies right (in case a header includes another header)?

I think the right way to do this is:

1. Add a `_headers` field to `GoArchive` in `emit_archive`. It should be a `depset` of `File`. The `direct` field should be headers from the current package. The `transitive` field should be the `GoArchive._headers` depsets from dependencies.
2. Pass that set through to `emit_compilepkg` as a new `headers` parameter. It's a `depset`, so just add it to the `inputs_transitive` list there.

Reply:
I _think_ I addressed this issue. I still added `_headers` to `GoArchiveData` in order to collect transitive headers.

Reply:
Please add to `GoArchive` instead.

`GoArchiveData` holds information about one specific package. It shouldn't contain information about dependencies. Instead, `GoArchive` provides access to all transitively imported packages through the `transitive` field, which is a depset of `GoArchiveData`.

Normally I'd suggest reconstructing the headers depset in each package by iterating over all the `GoArchiveData` accessible from `GoArchive`, but that makes analysis at least O(n^2), maybe worse, even in the common case where there are no headers. There's probably no avoiding O(n^2) here, but perhaps we can avoid most of that by adding a `_headers` field that only contains the headers and is empty most of the time.

Reply:
Maybe I'm doing something wrong but adding to `GoArchive` is causing infinite recompiles of stdlib.

Reply:
I think I found my mistake

Reply:
Updated with `headers` added to `GoArchive`. Happy to rename to `_headers` if you feel strongly.

Reply:
Please do rename to `_headers`. It communicates that it's a private field. I don't think other rules need to depend on it.

Reply:
done